db/types/operations/cast.py

```diff
@@ -256,6 +267,41 @@ def _escape_illegal_characters(sql_name):
     return resulting_string
 
 
+def _get_json_type_body_map(target_type):
+    """
+    json --> jsonb(or jsonb --> json): use default casting
+    text --> json: use _get_text_to_json_cast()
+    """
+    default_behavior_source_types = frozenset([PostgresType.JSON, PostgresType.JSONB])
+    source_text_types = categories.STRING_TYPES
+    target_type_str = target_type.id
+
+    def _get_text_to_json_cast(target_type):
+        return f"""
+                DECLARE json_res {target_type_str};
+                BEGIN
+                    SELECT
+                        CASE WHEN $1 IS NULL THEN NULL
+                        ELSE $1::{target_type_str}
+                        END
+                        INTO json_res;
+                IF {target_type_str}_typeof(json_res) != 'object' AND  {target_type_str}_typeof(json_res) != 'array'
+                    THEN RAISE EXCEPTION 'Invalid json expression';
+                END IF;
```

Comment:
This condition does not handle json null values. If the value of `json_res` is `"null"` (This is not the SQL NULL, this is json null value), then this condition would throw an exception. We should probably allow json null values and convert them to SQL NULLs.

@mathemancer Thoughts on this?

Reply:
I'm not sure whether a bare JSON null is even possible, since both SQL `NULL` and JSON `null` are represented as `None` in Python.